xilinx_test/host/RunHBMandBlockCopy.cpp

```diff
@@ -0,0 +1,168 @@
+    std::cout << "Done." << std::endl;
+
+    std::cout << "Initializing memory..." << std::flush;
+    std::array<size_t, 3> buf1Size = {5, 5, 5};
+    std::array<size_t, 3> buf2Size = {3, 3, 3};
+    size_t buf1Elems = buf1Size[0]*buf1Size[1]*buf1Size[2];
+    size_t buf2Elems = buf2Size[0]*buf2Size[1]*buf2Size[2];
+    std::vector<double, hlslib::ocl::AlignedAllocator<double, 4096>> memHostBuf1(buf1Elems);
+    //in sw_emu xilinx seems to just take over the ranges provided to it as device memory (during construction), 
+    //so don't use memDeviceBuf2
+    std::vector<double, hlslib::ocl::AlignedAllocator<double, 4096>> memDeviceBuf2(buf2Elems);
+    std::vector<double, hlslib::ocl::AlignedAllocator<double, 4096>> memHostBuf2(buf2Elems);
+    std::fill(memDeviceBuf2.begin(), memDeviceBuf2.end(), 3.0);
+    std::fill(memHostBuf1.begin(), memHostBuf1.end(), 1.0);
+    std::fill(memHostBuf2.begin(), memHostBuf2.end(), 2.0);
+
+    //A program needs to be loaded because Xilinx does not initialize the buffers otherwise
+    if (argc < 2) {
+        std::cerr << "Required argument: xclbin-file. It can be any xclbin file, since the kernel is not actually"
+        << "executed, but if no program is loaded, vitis will not create the buffers" << std::endl << std::flush;
+        return 1;
+  }
+
+    auto program = context.MakeProgram(argv[1]); 
+
+    auto memDevice1 = context.MakeBuffer<double, hlslib::ocl::Access::readWrite>(
+        hlslib::ocl::StorageType::DDR, 1, buf1Elems);
+    auto memDevice2 = context.MakeBuffer<double, hlslib::ocl::Access::readWrite>(
+        hlslib::ocl::StorageType::HBM, 10, memDeviceBuf2.begin(), memDeviceBuf2.end());
+    std::cout << " Done." << std::endl << std::flush;
+
+    std::cout << "Copy data around" << std::endl << std::flush;
+    double* dptr = nullptr;
+
+    //Some checks for the simulation. The check function is not used here, because it's based on the function tested here;
+    //In other words the assertions here are more to test the check function than the Copy operations
+    memDevice1.CopyBlockFromHost({0, 0, 0}, {0, 0, 0}, {5, 5, 5}, buf1Size, buf1Size, memHostBuf1.begin());
+    #ifdef HLSLIB_SIMULATE_OPENCL
+    dptr = memDevice1.devicePtr();
+    for(int i = 0; i < 125; i++) {
+        if(dptr[i] != 1) {
+            return 1;
+        }
+    }
+    assert(checkBlockHasValue({0, 0, 0}, {5, 5, 5}, buf1Size, dptr, 1.0));
+    #endif
+    memDevice2.CopyBlockToDevice({1, 1, 1}, {1, 1, 1}, {2, 2, 2}, buf2Size, buf1Size, memDevice1);
+    #ifdef HLSLIB_SIMULATE_OPENCL
+    dptr = memDevice1.devicePtr();
+    for(int i = 0; i < 25; i++) {
+        if(dptr[i] != 1) {
+            return 1;
+        }
+    }
+    for(int i = 25+5+1; i < 25+5+1+2; i++) {
+        if(dptr[i] != 3) {
+            return 1;
+        }
+    }
+    assert(checkBlockHasValue({1, 1, 1}, {2, 2, 2}, buf1Size, dptr, 3.0));
+    assert(checkBlockHasValue({0, 0, 3}, {5, 5, 2}, buf1Size, dptr, 1.0));
+    #endif
```

Comment:
Hmm, does that mean that this test only works if `HLSLIB_SIMULATE_OPENCL` is set, or just that we can't test this specific part? It sounds like we should compile two versions of this test?

Reply:
Added a second Test for this